### PR TITLE
[Homework 2] CRUD operations in memory

### PR DESCRIPTION
**Pull request by VictorBelikov**
# TASK 1
* To create REST service, use ExpressJS (https://expressjs.com/).
The User should have the following properties (you can use UUID as a user identifier (id)):
```
type User = {
  id: string;
  login: string;
  password: string;
  age: number;
  isDeleted: boolean;
};
````
* service should have the following CRUD operations for User:
− get user by id;
− create and update user;
− get auto-suggest list from limit users, sorted by login property and filtered by
loginSubstring in the login property:
getAutoSuggestUsers(loginSubstring, limit)
− remove user (soft delete – user gets marked with isDeleted flag, but not removed from
the collection).
* Store user’s collection in the service memory (while the service is running).
# TASK 2
Add server-side validation for create/update operations of User entity:
• all fields are required;
• login validation is required;
• password must contain letters and numbers;
• user’s age must be between 4 and 130.
In case of any property does not meet the validation requirements or the field is absent, return 400
(Bad Request) and detailed error message.
For requests validation use special packages like joi
(https://github.com/hapijs/joi, https://www.npmjs.com/package/express-joi-validation).

### CODE REVIEW

File: src/api/controllers/user.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    if (!err.statusCode) {
+      err.statusCode = 500;
+    }
+    return next(err);
+  }
+};
+
+export const createUser = (req, res, next) => {
+  try {
+    const newUser = new User(req.body.login, req.body.password, req.body.age).save();
```
Comment (VitaliBrych333) [inline src/api/controllers/user.js]: use destructuring

File: src/api/controllers/user.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+export const updateUser = (req, res, next) => {
+  try {
+    const user = User.findById(+req.params.userId);
+    if (!user) {
+      const error = new Error('Could not find a user!');
+      error.statusCode = 404;
+      throw error;
+    }
+    const updatedUser = User.update(+req.params.userId, new User(req.body.login, req.body.password, req.body.age));
```
Comment (VitaliBrych333) [inline src/api/controllers/user.js]: use destructuring

File: src/app.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+// Logs request type to the console
+app.use(morgan('dev'));
+
+// Body parser
+app.use(express.urlencoded({ extended: false }));
+app.use(express.json());
+
+// Send 'headers' to overcome CORS
+app.use((req, res, next) => {
+  res.setHeader('Access-Control-Allow-Origin', '*');
```
Comment (VitaliBrych333) [inline src/app.js]: What reason to allow CORS requests from any(*) resources?

File: src/api/controllers/user.js (hunk 1 above)
Comment (VictorBelikov) [reply src/api/controllers/user.js]: Done

File: src/api/controllers/user.js (hunk 2 above)
Comment (VictorBelikov) [reply src/api/controllers/user.js]: Done

File: src/app.js (hunk 3 above)
Comment (VictorBelikov) [reply src/app.js]: Removed this part of the code

**Review by VitaliBrych333 — APPROVED** [review]

**Pull request merged**